File: Leetcode/0543/543_Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <math.h>

class Solution {
public:
    int res = INT_MIN;
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == nullptr)
        {
            return 0;
        }
        
        int left = findMaxDiameter(root->left);
        int right = findMaxDiameter(root->right);
        
        if(res < left + right)
        {
            res = left + right;
        }
        
        diameterOfBinaryTree(root->left);
        diameterOfBinaryTree(root->right);
        
        return res;
    }
    
    int findMaxDiameter(TreeNode* root)
    {
        if(root == nullptr)
        {
            return 0;
        }
        int left = findMaxDiameter(root->left);
        int right = findMaxDiameter(root->right);
        
        if(left > right)
        {
            return left + 1;
        }
        else
        {
            return right + 1;
        }
    }
};

File: Leetcode/0543/543_Diameter_of_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    return traversal(root, 0);
};

var traversal = function(root, diameter) {
    if (root) {
        let rootDiameter = (root.left ? calculateMaxDepth(root.left, 1) : 0) + (root.right ? calculateMaxDepth(root.right, 1) : 0);
        
        diameter = rootDiameter > diameter ? rootDiameter : diameter; 
        const leftDiameter = traversal(root.left, diameter);
        const rightDiameter = traversal(root.right, diameter);
        
        const compare = leftDiameter > rightDiameter ? leftDiameter : rightDiameter;
        
        return compare > diameter ? compare : diameter;
    }
    return diameter;
}

var calculateMaxDepth = function(root, depth) {
    if (root) {
        const right = root.right ? calculateMaxDepth(root.right, depth + 1) : depth;
        const left = root.left ? calculateMaxDepth(root.left, depth + 1) : depth;
        
        return right > left ? right : left;
    }
    return depth;
}